<h1><center>Module 17: Challenge</center></h1>

### Assignment:

1) Load the houseprices data from Thinkful's database.

2) Do data cleaning, exploratory data analysis, and feature engineering. You can use your previous work in this module. But make sure that your work is satisfactory.

3) Now, split your data into train and test sets where 20% of the data resides in the test set.

4) Build several linear regression models including Lasso, Ridge, or ElasticNet and train them in the training set. Use k-fold cross-validation to select the best hyperparameters if your models include one!

5) Evaluate your best model on the test set.

6) So far, you have only used the features in the dataset. However, house prices can be affected by many factors like economic activity and the interest rates at the time they are sold. So, try to find some useful factors that are not included in the dataset. Integrate these factors into your model and assess the prediction performance of your model. Discuss the implications of adding these external variables into your model.

_________________________________________________________________________________________________________________

##### Import packages

In [1]:
import numpy as np
import pandas as pd
from sklearn import linear_model
import matplotlib.pyplot as plt
from sqlalchemy import create_engine
import statsmodels.api as sm
import matplotlib.pyplot as plt
import seaborn as sns
from statsmodels.tsa.stattools import acf
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error
from statsmodels.tools.eval_measures import mse, rmse
from sklearn.linear_model import LinearRegression, LassoCV, RidgeCV, ElasticNetCV


import warnings
warnings.filterwarnings('ignore')

pd.set_option('display.max_columns', 500)

##### Step 1: Load the houseprices data from Thinkful's database

In [2]:
postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'houseprices'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

houseprices_df = pd.read_sql_query('select * from houseprices',con=engine)

# no need for an open connection, 
# as we're only doing a single query
engine.dispose()

In [3]:
houseprices_df.head()

id  mssubclass mszoning  lotfrontage  lotarea street alley lotshape  \
0   1          60       RL         65.0     8450   Pave  None      Reg   
1   2          20       RL         80.0     9600   Pave  None      Reg   
2   3          60       RL         68.0    11250   Pave  None      IR1   
3   4          70       RL         60.0     9550   Pave  None      IR1   
4   5          60       RL         84.0    14260   Pave  None      IR1   

  landcontour utilities lotconfig landslope neighborhood condition1  \
0         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1         Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3         Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4         Lvl    AllPub       FR2       Gtl      NoRidge       Norm   

  condition2 bldgtype housestyle  overallqual  overallcond  yearbuilt  \
0       Norm     1Fam     2Story            7            5       2003   
1       Norm     1Fam     1Story            6            8       1976   
2       Norm     1Fam     2Story            7            5       2001   
3       Norm     1Fam     2Story            7            5       1915   
4       Norm     1Fam     2Story            8            5       2000   

   yearremodadd roofstyle roofmatl exterior1st exterior2nd masvnrtype  \
0          2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1          1976     Gable  CompShg     MetalSd     MetalSd       None   
2          2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
3          1970     Gable  CompShg     Wd Sdng     Wd Shng       None   
4          2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   

   masvnrarea exterqual extercond foundation bsmtqual bsmtcond bsmtexposure  \
0       196.0        Gd        TA      PConc       Gd       TA           No   
1         0.0        TA        TA     CBlock       Gd       TA           Gd   
2       162.0        Gd        TA      PConc       Gd       TA           Mn   
3         0.0        TA        TA     BrkTil       TA       Gd           No   
4       350.0        Gd        TA      PConc       Gd       TA           Av   

  bsmtfintype1  bsmtfinsf1 bsmtfintype2  bsmtfinsf2  bsmtunfsf  totalbsmtsf  \
0          GLQ         706          Unf           0        150          856   
1          ALQ         978          Unf           0        284         1262   
2          GLQ         486          Unf           0        434          920   
3          ALQ         216          Unf           0        540          756   
4          GLQ         655          Unf           0        490         1145   

  heating heatingqc centralair electrical  firstflrsf  secondflrsf  \
0    GasA        Ex          Y      SBrkr         856          854   
1    GasA        Ex          Y      SBrkr        1262            0   
2    GasA        Ex          Y      SBrkr         920          866   
3    GasA        Gd          Y      SBrkr         961          756   
4    GasA        Ex          Y      SBrkr        1145         1053   

   lowqualfinsf  grlivarea  bsmtfullbath  bsmthalfbath  fullbath  halfbath  \
0             0       1710             1             0         2         1   
1             0       1262             0             1         2         0   
2             0       1786             1             0         2         1   
3             0       1717             1             0         1         0   
4             0       2198             1             0         2         1   

   bedroomabvgr  kitchenabvgr kitchenqual  totrmsabvgrd functional  \
0             3             1          Gd             8        Typ   
1             3             1          TA             6        Typ   
2             3             1          Gd             6        Typ   
3             3             1          Gd             7        Typ   
4             4             1          Gd             9        

In [4]:
houseprices_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
id               1460 non-null int64
mssubclass       1460 non-null int64
mszoning         1460 non-null object
lotfrontage      1201 non-null float64
lotarea          1460 non-null int64
street           1460 non-null object
alley            91 non-null object
lotshape         1460 non-null object
landcontour      1460 non-null object
utilities        1460 non-null object
lotconfig        1460 non-null object
landslope        1460 non-null object
neighborhood     1460 non-null object
condition1       1460 non-null object
condition2       1460 non-null object
bldgtype         1460 non-null object
housestyle       1460 non-null object
overallqual      1460 non-null int64
overallcond      1460 non-null int64
yearbuilt        1460 non-null int64
yearremodadd     1460 non-null int64
roofstyle        1460 non-null object
roofmatl         1460 non-null object
exterior1st      1460 non-n

##### Step 2: Do data cleaning, exploratory data analysis, and feature engineering.

In [5]:
houseprices_df.describe()

id   mssubclass  lotfrontage        lotarea  overallqual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       overallcond    yearbuilt  yearremodadd   masvnrarea   bsmtfinsf1  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726   
std       1.112799    30.202904     20.645407   181.066207   456.098091   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000   

        bsmtfinsf2    bsmtunfsf  totalbsmtsf   firstflrsf  secondflrsf  \
count  1460.000000  1460.000000  1460.000000  1460.000000  1460.000000   
mean     46.549315   567.240411  1057.429452  1162.626712   346.992466   
std     161.319273   441.866955   438.705324   386.587738   436.528436   
min       0.000000     0.000000     0.000000   334.000000     0.000000   
25%       0.000000   223.000000   795.750000   882.000000     0.000000   
50%       0.000000   477.500000   991.500000  1087.000000     0.000000   
75%       0.000000   808.000000  1298.250000  1391.250000   728.000000   
max    1474.000000  2336.000000  6110.000000  4692.000000  2065.000000   

       lowqualfinsf    grlivarea  bsmtfullbath  bsmthalfbath     fullbath  \
count   1460.000000  1460.000000   1460.000000   1460.000000  1460.000000   
mean       5.844521  1515.463699      0.425342      0.057534     1.565068   
std       48.623081   525.480383      0.518911      0.238753     0.550916   
min        0.000000   334.000000      0.000000      0.000000     0.000000   
25%        0.000000  1129.500000      0.000000      0.000000     1.000000   
50%        0.000000  1464.000000      0.000000      0.000000     2.000000   
75%        0.000000  1776.750000      1.000000      0.000000     2.000000   
max      572.000000  5642.000000      3.000000      2.000000     3.000000   

          halfbath  bedroomabvgr  kitchenabvgr  totrmsabvgrd   fireplaces  \
count  1460.000000   1460.000000   1460.000000   1460.000000  1460.000000   
mean      0.382877      2.866438      1.046575      6.517808     0.613014   
std       0.502885      0.815778      0.220338      1.625393     0.644666   
min       0.000000      0.000000      0.000000      2.000000     0.000000   
25%       0.000000      2.000000      1.000000      5.000000     0.000000   
50%       0.000000      3.000000      1.000000      6.000000     1.000000   
75%       1.000000      3.000000      1.000000      7.000000     1.000000   
max       2.000000      8.000000      3.000000     14.000000     3.000000   

       garageyrblt   garagecars   garagearea   wooddecksf  openporchsf  \
count  1379.000000  1460.000000  1460.000000  1460.000000  1460.000000   
mean   1978.506164     1.767123   472.980137    94.244521    46.660274   
std      24.689725     0.747315   213.804841   125.338794    66.256028   
min    1900.000000     0.000000     0.000000     0.000000     0.000000   
25%    1961.000000     1.000000   334.500000     0.000000     0.000000   
50%    1980.000000     2.000000   480.000000     0.000000    25.000000   
75%    2002.000000     2.000000   576.000000   168.000000    68.000000   
max    2

>There are some columns that are numerical and some that are categorical. If we want to use the categorical then we will have to create dummy variables.

>create a list of numeric and categorical variables to see how many there are,

In [6]:
#create a list of numeric variables save as numeric_variables...
numeric_variables = houseprices_df.select_dtypes(['int64', 'float64']).columns
print(numeric_variables)
print("The number of numerical columns is {}".format(len(numeric_variables)))

Index(['id', 'mssubclass', 'lotfrontage', 'lotarea', 'overallqual',
       'overallcond', 'yearbuilt', 'yearremodadd', 'masvnrarea', 'bsmtfinsf1',
       'bsmtfinsf2', 'bsmtunfsf', 'totalbsmtsf', 'firstflrsf', 'secondflrsf',
       'lowqualfinsf', 'grlivarea', 'bsmtfullbath', 'bsmthalfbath', 'fullbath',
       'halfbath', 'bedroomabvgr', 'kitchenabvgr', 'totrmsabvgrd',
       'fireplaces', 'garageyrblt', 'garagecars', 'garagearea', 'wooddecksf',
       'openporchsf', 'enclosedporch', 'threessnporch', 'screenporch',
       'poolarea', 'miscval', 'mosold', 'yrsold', 'saleprice'],
      dtype='object')
The number of numerical columns is 38


In [7]:
#create a list of categorical variables save as categorical_variables...
categorical_variables = houseprices_df.select_dtypes(['object']).columns
print(categorical_variables)
print("The number of categorical columns is {}".format(len(categorical_variables)))

Index(['mszoning', 'street', 'alley', 'lotshape', 'landcontour', 'utilities',
       'lotconfig', 'landslope', 'neighborhood', 'condition1', 'condition2',
       'bldgtype', 'housestyle', 'roofstyle', 'roofmatl', 'exterior1st',
       'exterior2nd', 'masvnrtype', 'exterqual', 'extercond', 'foundation',
       'bsmtqual', 'bsmtcond', 'bsmtexposure', 'bsmtfintype1', 'bsmtfintype2',
       'heating', 'heatingqc', 'centralair', 'electrical', 'kitchenqual',
       'functional', 'fireplacequ', 'garagetype', 'garagefinish', 'garagequal',
       'garagecond', 'paveddrive', 'poolqc', 'fence', 'miscfeature',
       'saletype', 'salecondition'],
      dtype='object')
The number of categorical columns is 43


In [8]:
#find percent of missing values...
missing_percent = (houseprices_df.isnull().sum()/houseprices_df.isnull().count()).sort_values(ascending=False)
missing_percent.head(40)

poolqc          0.995205
miscfeature     0.963014
alley           0.937671
fence           0.807534
fireplacequ     0.472603
lotfrontage     0.177397
garagecond      0.055479
garagetype      0.055479
garageyrblt     0.055479
garagefinish    0.055479
garagequal      0.055479
bsmtexposure    0.026027
bsmtfintype2    0.026027
bsmtfintype1    0.025342
bsmtcond        0.025342
bsmtqual        0.025342
masvnrarea      0.005479
masvnrtype      0.005479
electrical      0.000685
utilities       0.000000
yearremodadd    0.000000
mssubclass      0.000000
foundation      0.000000
extercond       0.000000
exterqual       0.000000
exterior2nd     0.000000
exterior1st     0.000000
roofmatl        0.000000
roofstyle       0.000000
yearbuilt       0.000000
lotconfig       0.000000
overallcond     0.000000
overallqual     0.000000
housestyle      0.000000
bldgtype        0.000000
condition2      0.000000
bsmtfinsf1      0.000000
mszoning        0.000000
lotarea         0.000000
street          0.000000


##### Step 3: Split your data into train and test sets where 20% of the data resides in the test set.

In [9]:
# Y is the target variable
Y = houseprices_df['saleprice']
# X is the feature set
X = houseprices_df[['overallqual', 'grlivarea', 'garagecars', 'garagearea', 'totalbsmtsf']]

In [10]:
X = sm.add_constant(X)

X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size = 0.2, random_state = 465)

##### Step 4: Build several linear regression models including Lasso, Ridge, or ElasticNet and train them in the training set. Use k-fold cross-validation to select the best hyperparameters if your models include one!

* Linear Regression

In [11]:
lrm = LinearRegression()

lrm.fit(X_train, y_train)

# We are making predictions here
y_preds_train = lrm.predict(X_train)
y_preds_test = lrm.predict(X_test)

print("R-squared of the model in training set is: {}".format(lrm.score(X_train, y_train)))
print("-----Test set statistics-----")
print("R-squared of the model in test set is: {}".format(lrm.score(X_test, y_test)))
print("Mean absolute error of the prediction is: {}".format(mean_absolute_error(y_test, y_preds_test)))
print("Mean squared error of the prediction is: {}".format(mse(y_test, y_preds_test)))
print("Root mean squared error of the prediction is: {}".format(rmse(y_test, y_preds_test)))
print("Mean absolute percentage error of the prediction is: {}".format(np.mean(np.abs((y_test - y_preds_test) / y_test)) * 100))

R-squared of the model in training set is: 0.7589588843157826
-----Test set statistics-----
R-squared of the model in test set is: 0.7656379370075204
Mean absolute error of the prediction is: 25964.89937383237
Mean squared error of the prediction is: 1573437538.2926722
Root mean squared error of the prediction is: 39666.57961423788
Mean absolute percentage error of the prediction is: 16.09717844470052


* Lasso Regression

In [12]:
alphas = [np.power(10.0,p) for p in np.arange(-10,40,1)]

In [13]:

lasso_cv = LassoCV(alphas=alphas, cv=5)

lasso_cv.fit(X_train, y_train)

# We are making predictions here
y_preds_train = lasso_cv.predict(X_train)
y_preds_test = lasso_cv.predict(X_test)

print("Best alpha value is: {}".format(lasso_cv.alpha_))
print("R-squared of the model in training set is: {}".format(lasso_cv.score(X_train, y_train)))
print("-----Test set statistics-----")
print("R-squared of the model in test set is: {}".format(lasso_cv.score(X_test, y_test)))
print("Mean absolute error of the prediction is: {}".format(mean_absolute_error(y_test, y_preds_test)))
print("Mean squared error of the prediction is: {}".format(mse(y_test, y_preds_test)))
print("Root mean squared error of the prediction is: {}".format(rmse(y_test, y_preds_test)))
print("Mean absolute percentage error of the prediction is: {}".format(np.mean(np.abs((y_test - y_preds_test) / y_test)) * 100))

Best alpha value is: 1e-10
R-squared of the model in training set is: 0.7589588843157826
-----Test set statistics-----
R-squared of the model in test set is: 0.7656379370075171
Mean absolute error of the prediction is: 25964.899373832293
Mean squared error of the prediction is: 1573437538.2926934
Root mean squared error of the prediction is: 39666.579614238144
Mean absolute percentage error of the prediction is: 16.097178444700386


* Ridge Regression

In [14]:
ridge_cv = RidgeCV(alphas=alphas, cv=5)

ridge_cv.fit(X_train, y_train)

# We are making predictions here
y_preds_train = ridge_cv.predict(X_train)
y_preds_test = ridge_cv.predict(X_test)

print("Best alpha value is: {}".format(ridge_cv.alpha_))
print("R-squared of the model in training set is: {}".format(ridge_cv.score(X_train, y_train)))
print("-----Test set statistics-----")
print("R-squared of the model in test set is: {}".format(ridge_cv.score(X_test, y_test)))
print("Mean absolute error of the prediction is: {}".format(mean_absolute_error(y_test, y_preds_test)))
print("Mean squared error of the prediction is: {}".format(mse(y_test, y_preds_test)))
print("Root mean squared error of the prediction is: {}".format(rmse(y_test, y_preds_test)))
print("Mean absolute percentage error of the prediction is: {}".format(np.mean(np.abs((y_test - y_preds_test) / y_test)) * 100))

Best alpha value is: 10.0
R-squared of the model in training set is: 0.7589360914817471
-----Test set statistics-----
R-squared of the model in test set is: 0.7660352716902978
Mean absolute error of the prediction is: 25888.340565636823
Mean squared error of the prediction is: 1570769950.8121517
Root mean squared error of the prediction is: 39632.94022416393
Mean absolute percentage error of the prediction is: 16.038387948827452


* Elastic Net Regression

In [15]:

elasticnet_cv = ElasticNetCV(alphas=alphas, cv=5)

elasticnet_cv.fit(X_train, y_train)

# We are making predictions here
y_preds_train = elasticnet_cv.predict(X_train)
y_preds_test = elasticnet_cv.predict(X_test)

print("Best alpha value is: {}".format(elasticnet_cv.alpha_))
print("R-squared of the model in training set is: {}".format(elasticnet_cv.score(X_train, y_train)))
print("-----Test set statistics-----")
print("R-squared of the model in test set is: {}".format(elasticnet_cv.score(X_test, y_test)))
print("Mean absolute error of the prediction is: {}".format(mean_absolute_error(y_test, y_preds_test)))
print("Mean squared error of the prediction is: {}".format(mse(y_test, y_preds_test)))
print("Root mean squared error of the prediction is: {}".format(rmse(y_test, y_preds_test)))
print("Mean absolute percentage error of the prediction is: {}".format(np.mean(np.abs((y_test - y_preds_test) / y_test)) * 100))

Best alpha value is: 0.01
R-squared of the model in training set is: 0.7589507685767886
-----Test set statistics-----
R-squared of the model in test set is: 0.7658770134292637
Mean absolute error of the prediction is: 25918.68116349639
Mean squared error of the prediction is: 1571832449.944803
Root mean squared error of the prediction is: 39646.34220132802
Mean absolute percentage error of the prediction is: 16.062121890304276


##### Step 5: Evaluate your best model on the test set.

To evaluate which model has the best performance, I will compare the R-squared values. The R-squared values are as follows:

* Linear Regression - R-squared of the model in test set is: 0.7656379370075204
* Lasso Regression - R-squared of the model in training set is: 0.7589588843157826
* Ridge Regression - R-squared of the model in training set is: 0.7589360914817471
* Elastic Net Regression - R-squared of the model in test set is: 0.7658770134292637

The Elastic Net Regression has the highest overall R-squared value and is therefore the most effective model.

##### Step 6: Try to find some useful factors that are not included in the dataset. Integrate these factors into your model and assess the prediction performance of your model. Discuss the implications of adding these external variables into your model.

An additional factor that I would consider adding into my model to determine house price would be the overall unemployment rate at the time a house is sold. If the unemployment rate is high then I would expect for their to be a negative effect on the final sale price of a house in a given time.

##### Add in the unemployment rate

In [59]:
#read in the unemployment data...
unemployment_df = pd.read_excel('UNRATE.xls', header = 10)

In [60]:
unemployment_df.head()

observation_date  UNRATE
0       2006-01-01     4.7
1       2006-02-01     4.8
2       2006-03-01     4.7
3       2006-04-01     4.7
4       2006-05-01     4.6

##### Start to separate the observation rate and create an 'id' variable that we will then be able to merge on.

In [61]:
#create a year variable...
unemployment_df['year'] = unemployment_df['observation_date'].astype(str).str[0:4]

In [62]:
#create a month variable...
unemployment_df['month'] = unemployment_df['observation_date'].astype(str).str[5:7]

In [63]:
#replace the zero in the month column with nothing...
unemployment_df['month'] = unemployment_df['month'].str.replace("0","")

In [64]:
#create a unique id variable to merge on later...
unemployment_df['id'] = unemployment_df['year'] + "_" + unemployment_df['month']

In [29]:
unemployment_df.head()

observation_date  UNRATE  year month      id
0       2006-01-01     4.7  2006     1  2006_1
1       2006-02-01     4.8  2006     2  2006_2
2       2006-03-01     4.7  2006     3  2006_3
3       2006-04-01     4.7  2006     4  2006_4
4       2006-05-01     4.6  2006     5  2006_5

##### Create 'id' variable in houseprices to merge on

In [65]:
#turn variables into a string and create new id variable...
houseprices_df['yrsold'] = houseprices_df['yrsold'].astype('str')
houseprices_df['mosold'] = houseprices_df['mosold'].astype('str')
houseprices_df['id'] = houseprices_df['yrsold'] + "_" + houseprices_df['mosold']

In [66]:
#merge on 'id' created...
houseprices_unemployment_df = pd.merge(houseprices_df, unemployment_df, how='inner', on=['id'])

In [58]:
result.head(100)

id  mssubclass mszoning  lotfrontage  lotarea street alley lotshape  \
0   2008_2          60       RL         65.0     8450   Pave  None      Reg   
1   2008_2          20       RL         70.0    11200   Pave  None      Reg   
2   2008_2          20       RL         75.0    10125   Pave  None      Reg   
3   2008_2          30       RM         71.0     6900   Pave  None      Reg   
4   2008_2          20       RL        100.0    21750   Pave  None      Reg   
..     ...         ...      ...          ...      ...    ...   ...      ...   
95  2007_8         120       RL         60.0     7313   Pave  None      Reg   
96  2007_8          20       RL         60.0     8070   Pave  None      Reg   
97  2007_8          30       RM         60.0    10800   Pave  Grvl      Reg   
98  2007_8          50       RM         85.0     8500   Pave  None      Reg   
99  2007_8          30       RM         40.0     3880   Pave  None      Reg   

   landcontour utilities lotconfig landslope neighborhood condition1  \
0          Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1          Lvl    AllPub    Inside       Gtl       Sawyer       Norm   
2          Lvl    AllPub    Inside       Gtl      Mitchel       Norm   
3          Lvl    AllPub    Inside       Gtl       IDOTRR       Norm   
4          HLS    AllPub    Inside       Mod      Mitchel     Artery   
..         ...       ...       ...       ...          ...        ...   
95         Lvl    AllPub    Inside       Gtl      NridgHt       Norm   
96         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
97         Lvl    AllPub    Inside       Gtl      OldTown       Norm   
98         Lvl    AllPub    Corner       Gtl       IDOTRR     Artery   
99         Lvl    AllPub    Inside       Gtl      OldTown       Norm   

   condition2 bldgtype housestyle  overallqual  overallcond  yearbuilt  \
0        Norm     1Fam     2Story            7            5       2003   
1        Norm     1Fam     1Story            5            5       1965   
2        Norm     1Fam     1Story            6            6       1977   
3        Norm     1Fam     1Story            5            6       1940   
4        Norm     1Fam     1Story            5            5       1954   
..        ...      ...        ...          ...          ...        ...   
95       Norm   TwnhsE     1Story            9            5       2005   
96       Norm     1Fam     1Story            4            5       1994   
97       Norm     1Fam     1Story            4            7       1885   
98       Norm     1Fam     1.5Fin            5            7       1919   
99       Norm     1Fam     1Story            5            9       1945   

    yearremodadd roofstyle roofmatl exterior1st exterior2nd masvnrtype  \
0           2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1           1965       Hip  CompShg     HdBoard     HdBoard       None   
2           1977     Gable  CompShg     Plywood     Plywood       None   
3           1955     Gable  CompShg     VinylSd     VinylSd       None   
4           1954       Hip  CompShg     HdBoard     HdBoard       None   
..           ...       ...      ...         ...         ...        ...   
95          2005       Hip  CompShg     MetalSd     MetalSd    BrkFace   
96          1995     Gable  CompShg     VinylSd     VinylSd       None   
97          1995   Mansard  CompShg     VinylSd     VinylSd       None   
98          2005     Gable  CompShg     CemntBd     CmentBd       None   
99          1997     Gable  CompShg     VinylSd     VinylSd       None   

    masvnrarea exterqual extercond foundation bsmtqual bsmtcond bsmtexposure  \
0        196.0        Gd        TA      PConc       Gd       TA           No   
1          0.0        TA        TA     CBlock       TA       TA           No   
2          0.0        TA        TA     CBlock       TA       TA           No   
3          0.0        TA        TA     CBlock       TA       TA           No   
4          0.0  

In [68]:
# Y is the target variable
Y = houseprices_unemployment_df['saleprice']
# X is the feature set
X = houseprices_unemployment_df[['overallqual', 'grlivarea', 'garagecars', 'garagearea', 'totalbsmtsf', 'UNRATE']]

In [69]:
X = sm.add_constant(X)

X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size = 0.2, random_state = 465)

In [70]:
lrm = LinearRegression()

lrm.fit(X_train, y_train)

# We are making predictions here
y_preds_train = lrm.predict(X_train)
y_preds_test = lrm.predict(X_test)

print("R-squared of the model in training set is: {}".format(lrm.score(X_train, y_train)))
print("-----Test set statistics-----")
print("R-squared of the model in test set is: {}".format(lrm.score(X_test, y_test)))
print("Mean absolute error of the prediction is: {}".format(mean_absolute_error(y_test, y_preds_test)))
print("Mean squared error of the prediction is: {}".format(mse(y_test, y_preds_test)))
print("Root mean squared error of the prediction is: {}".format(rmse(y_test, y_preds_test)))
print("Mean absolute percentage error of the prediction is: {}".format(np.mean(np.abs((y_test - y_preds_test) / y_test)) * 100))

R-squared of the model in training set is: 0.7656257240452164
-----Test set statistics-----
R-squared of the model in test set is: 0.6833720515562414
Mean absolute error of the prediction is: 28153.606886986956
Mean squared error of the prediction is: 2512576747.282546
Root mean squared error of the prediction is: 50125.609694871004
Mean absolute percentage error of the prediction is: 15.96550317144202


> Comparing the model effectiveness with the unemployment rate added in and without the unemployment rate added in, we see that without the unemployment rate we have an R-squared of 0.7589 and with the unemployment rate added in we have an R-squared of 0.7656. Our R-squared is slightly higher when we had in the unemployment so we can therefore say that adding in this variable increases our model effectiveness. However, it does not increase the R-squared value substantially and therefore does not substantially improve our model effectiveness.